### PR TITLE
[Hackthon No.26]

### PR DESCRIPTION
**Pull request by yangguohao**

### CODE REVIEW

**Comment by dingjiaweiww** [comment]
你的 PR 提交成功，感谢你对于开源项目的贡献，请检查 PR 提交格式和内容是否完备，具体请参考[示例](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20200301_api_design_for_quantile.md)和[模版](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/api_design_template.md)。

**Comment by dingjiaweiww** [comment]
PR 格式检查通过，你的PR 将接受Paddle 专家以及开源社区的review，请及时关注PR 动态

File: rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+> $$
+>
+> where
+>
+> $$
+> d\left(x_{i}, y_{i}\right)=\left\|\mathbf{x}_{i}-\mathbf{y}_{i}\right\|_{p}
+> $$
+
+# 四、对比分析
+paddle 和 pytorch整体框架相似，故直接采用其方法进行设计。
```
[the remark this answers is not in the document]
Comment (shiyutang) [reply rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md:47]: 即两种方案的优劣

File: rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+# 五、设计思路与实现方案
+
+## 命名与参数设计
+参考：[飞桨API 设计及命名规范]
```
Comment (shiyutang) [inline rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md:52]: 说明可以删除

File: rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+共添加以下两个 API：
+
+- `paddle.nn.TripletMarginLoss(margin=1.0, p=2.0, epsilon=1e-06, swap=False, reduction='mean', name=None) -> Tensor`
+- `padde.nn.functional.triplet_margin_loss(input, positive, negative, margin=1.0, p=2, epsilon=1e-06, swap=False, reduction='mean', name=None) -> Tensor`
```
Comment (shiyutang) [inline rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md:56]: 输入和参数支持的维度、参数范围可以进一步说明

File: rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+## 底层OP设计
+## API实现方案
+distance functions可以采用paddle.nn.PairWiseDistance来进行实现
+1. 检查参数
+
+   1. 检查 reduction 有效性（同其余 functional loss 中的实现）
+   2. 检查输入的 dtype（含 `anchor`、`positive`、`negative`）（同其余 functional loss 中的实现）
+
```
Comment (shiyutang) [inline rfcs/APIs/ 20220313_api_design_for_tripletmarginloss.md:64]: 3. 检查 `anchor`、`positive`、`negative` 三者维度是否对齐

**Comment by shiyutang** [comment]
如果修改完整，请rerequest-review
image

File: rfcs/APIs/20220313_api_design_for_tripletmarginloss.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+distance functions可以采用paddle.nn.PairWiseDistance来进行实现
+1. 检查参数
+
+   1. 检查 reduction 有效性（同其余 functional loss 中的实现）
+   2. 检查输入的 dtype（含 `input`、`positive`、`negative`）（同其余 functional loss 中的实现）
+   3. 检查参数维度是否相同
+ 
+2. 计算
+
+   1. 使用 `paddle.nn.Pairwisedistance` 分别计算得到正锚点与样本和负锚点与样本的距离。
```
Comment (shiyutang) [inline rfcs/APIs/20220313_api_design_for_tripletmarginloss.md:181]: 可以替换为 paddle.linalg.norm 底层和nn.PairwiseDistance 调用相同的 pnorm OP，并且可以直接进行函数调用

File: rfcs/APIs/20220313_api_design_for_tripletmarginloss.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+
+# 五、设计思路与实现方案
+
+## 命名与参数设计
+共添加以下两个 API：
+
+```
+padde.nn.functional.triplet_margin_loss(input Tensor[float64 or float32] 维度为[batch_size,dim] 
+                                          positive, 1 or -1 Tensor[float64 or float32],维度为[batch_size,dim]                                                                                           negative, 1 or -1 Tensor[float64 or float32],维度为[batch_size,dim]
```
Comment (shiyutang) [inline rfcs/APIs/20220313_api_design_for_tripletmarginloss.md:157]: 为什么positive和negative需要限制为-1/1的tensor？

File: rfcs/APIs/20220313_api_design_for_tripletmarginloss.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                              reduction='mean', 
+                              name=None) -> Tensor
+```
+## 底层OP设计
+## API实现方案
+distance functions可以采用paddle.nn.PairWiseDistance来进行实现
+1. 检查参数
+
+   1. 检查 reduction 有效性（同其余 functional loss 中的实现）
+   2. 检查输入的 dtype（含 `input`、`positive`、`negative`）（同其余 functional loss 中的实现）
```
Comment (shiyutang) [inline rfcs/APIs/20220313_api_design_for_tripletmarginloss.md:176]: 由于输入损失的tensor不一定为（N，Dim）的形状，还需要对输入进行维度转换

**Review by shiyutang — CHANGES_REQUESTED** [review]
方案实现细节存在修改部分

**Review by shiyutang — APPROVED** [review]
LGTM
